File: android/app/src/main/java/com/buildapp/photos/api/PhotoApi.kt
package com.buildapp.photos.api

import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query
import java.util.concurrent.TimeUnit

interface PhotoApi {
    @GET("api/health")
    suspend fun health(): Health

    @GET("api/media")
    suspend fun media(
        @Query("page") page: Int,
        @Query("per_page") perPage: Int = 60,
        @Query("kind") kind: String? = null,
        @Query("album") album: String? = null,
        @Query("q") q: String? = null,
        @Query("favorites") favorites: Int? = null,
        @Query("year") year: Int? = null,
        @Query("month") month: Int? = null,
        @Query("trashed") trashed: Int? = null,
        @Query("archived") archived: Int? = null,
    ): MediaPage

    @GET("api/albums")
    suspend fun albums(): List<Album>

    @GET("api/timeline")
    suspend fun timeline(): List<TimelineBucket>

    @GET("api/memories")
    suspend fun memories(): Memories

    @GET("api/search_semantic")
    suspend fun searchSemantic(
        @Query("q") q: String,
        @Query("top_k") topK: Int = 80,
    ): MediaPage

    @GET("api/clusters")
    suspend fun clusters(): List<Cluster>

    @GET("api/clusters/{id}/media")
    suspend fun clusterMedia(@Path("id") id: Int): List<MediaItem>

    @GET("api/faces/status")
    suspend fun facesStatus(): FacesStatus

    @GET("api/persons")
    suspend fun persons(): List<Person>

    @POST("api/rescan")
    suspend fun rescan(): Map<String, String>

    @POST("api/media/{id}/favorite")
    suspend fun toggleFavorite(@Path("id") id: String): FavoriteResp

    @POST("api/media/{id}/trash")
    suspend fun trash(@Path("id") id: String): Map<String, kotlinx.serialization.json.JsonElement>

    @POST("api/media/{id}/restore")
    suspend fun restore(@Path("id") id: String): Map<String, kotlinx.serialization.json.JsonElement>

    @POST("api/media/{id}/archive")
    suspend fun archive(@Path("id") id: String): Map<String, kotlinx.serialization.json.JsonElement>

    @POST("api/media/{id}/rotate")
    suspend fun rotate(
        @Path("id") id: String,
        @Query("degrees") degrees: Int = 90,
    ): Map<String, kotlinx.serialization.json.JsonElement>

    companion object {
        fun create(baseUrl: String): PhotoApi {
            val json = Json {
                ignoreUnknownKeys = true
                explicitNulls = false
                coerceInputValues = true
            }
            val logging = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BASIC
            }
            val client = OkHttpClient.Builder()
                .addInterceptor(logging)
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .build()
            val baseNormalized = if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/"
            return Retrofit.Builder()
                .baseUrl(baseNormalized)
                .client(client)
                .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
                .build()
                .create(PhotoApi::class.java)
        }
    }
}

object Urls {
    fun thumb(baseUrl: String, id: String): String =
        (if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/") + "api/media/$id/thumb"

    fun full(baseUrl: String, id: String, asJpeg: Boolean = false): String {
        val base = if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/"
        return base + "api/media/$id/full" + if (asJpeg) "?as=jpeg" else ""
    }

    fun stream(baseUrl: String, id: String): String =
        (if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/") + "api/media/$id/stream.mp4"

    fun clusterThumb(baseUrl: String, id: Int): String =
        (if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/") + "api/clusters/$id/thumb"
}

File: ios/Chitra/API/PhotoAPI.swift
import Foundation

enum APIError: LocalizedError {
    case badURL(String)
    case http(Int)
    case decoding(String)

    var errorDescription: String? {
        switch self {
        case .badURL(let u): return "Bad server URL: \(u)"
        case .http(let code): return "HTTP \(code)"
        case .decoding(let what): return "Bad response (\(what))"
        }
    }

    /// The server refuses writes while the library is mounted read-only.
    var isForbidden: Bool {
        if case .http(403) = self { return true }
        return false
    }
}

/// The JSON API, one method per Retrofit endpoint in `PhotoApi.kt`.
struct PhotoAPI {
    let baseUrl: String
    private let session: URLSession

    init(baseUrl: String, session: URLSession? = nil) {
        self.baseUrl = Urls.base(baseUrl)
        if let session {
            self.session = session
        } else {
            let cfg = URLSessionConfiguration.default
            cfg.timeoutIntervalForRequest = 60
            cfg.waitsForConnectivity = false
            self.session = URLSession(configuration: cfg)
        }
    }

    private static let decoder = JSONDecoder()
    private static let encoder = JSONEncoder()

    // MARK: - Plumbing

    private func url(_ path: String, _ query: [String: String?] = [:]) throws -> URL {
        guard var comps = URLComponents(string: baseUrl + path) else { throw APIError.badURL(baseUrl + path) }
        let items = query.compactMap { key, value -> URLQueryItem? in
            guard let value else { return nil }
            return URLQueryItem(name: key, value: value)
        }.sorted { $0.name < $1.name }
        if !items.isEmpty { comps.queryItems = items }
        guard let u = comps.url else { throw APIError.badURL(baseUrl + path) }
        return u
    }

    private func send<T: Decodable>(_ request: URLRequest, as type: T.Type) async throws -> T {
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw APIError.http(http.statusCode)
        }
        do {
            return try Self.decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: T.self))
        }
    }

    private func get<T: Decodable>(_ path: String, _ query: [String: String?] = [:], as type: T.Type) async throws -> T {
        try await send(URLRequest(url: try url(path, query)), as: type)
    }

    private func request(_ method: String, _ path: String, _ query: [String: String?] = [:], body: Data? = nil) throws -> URLRequest {
        var r = URLRequest(url: try url(path, query))
        r.httpMethod = method
        if let body {
            r.httpBody = body
            r.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        return r
    }

    /// A write whose response body we don't care about beyond "it worked".
    @discardableResult
    private func fire(_ method: String, _ path: String, _ query: [String: String?] = [:], body: Data? = nil) async throws -> [String: JSONValue] {
        try await send(try request(method, path, query, body: body), as: [String: JSONValue].self)
    }

    // MARK: - Reads

    func health() async throws -> Health {
        try await get("api/health", as: Health.self)
    }

    func meta(_ id: String) async throws -> MediaItem {
        try await get("api/media/\(id)", as: MediaItem.self)
    }

    func media(page: Int, perPage: Int = 60, kind: String? = nil, album: String? = nil,
               folder: String? = nil, q: String? = nil, favorites: Int? = nil,
               year: Int? = nil, month: Int? = nil, trashed: Int? = nil,
               archived: Int? = nil, dated: Int? = nil, undated: Int? = nil,
               sort: String? = nil) async throws -> MediaPage {
        try await get("api/media", [
            "page": String(page), "per_page": String(perPage),
            "kind": kind, "album": album, "folder": folder, "q": q,
            "favorites": favorites.map(String.init), "year": year.map(String.init),
            "month": month.map(String.init), "trashed": trashed.map(String.init),
            "archived": archived.map(String.init), "dated": dated.map(String.init),
            "undated": undated.map(String.init), "sort": sort,
        ], as: MediaPage.self)
    }

    func albums() async throws -> [Album] {
        try await get("api/albums", as: [Album].self)
    }

    func userAlbums(mediaId: String? = nil) async throws -> [UserAlbum] {
        try await get("api/user_albums", ["media_id": mediaId], as: [UserAlbum].self)
    }

    func userAlbumMedia(_ id: Int) async throws -> [MediaItem] {
        try await get("api/user_albums/\(id)/media", as: [MediaItem].self)
    }

    func timeline() async throws -> [TimelineBucket] {
        try await get("api/timeline", as: [TimelineBucket].self)
    }

    func memories() async throws -> Memories {
        try await get("api/memories", as: Memories.self)
    }

    func searchSemantic(q: String, topK: Int = 80) async throws -> MediaPage {
        try await get("api/search_semantic", ["q": q, "top_k": String(topK)], as: MediaPage.self)
    }

    func clusters() async throws -> [Cluster] {
        try await get("api/clusters", as: [Cluster].self)
    }

    func clusterMedia(_ id: Int) async throws -> [MediaItem] {
        try await get("api/clusters/\(id)/media", as: [MediaItem].self)
    }

    func facesStatus() async throws -> FacesStatus {
        try await get("api/faces/status", as: FacesStatus.self)
    }

    func persons() async throws -> [Person] {
        try await get("api/persons", as: [Person].self)
    }

    func locations() async throws -> [LocationItem] {
        try await get("api/locations", as: [LocationItem].self)
    }

    // MARK: - Writes

    func rescan() async throws {
        try await fire("POST", "api/rescan")
    }

    func toggleFavorite(_ id: String) async throws -> FavoriteResp {
        try await send(try request("POST", "api/media/\(id)/favorite"), as: FavoriteResp.self)
    }

    func trash(_ id: String) async throws {
        try await fire("POST", "api/media/\(id)/trash")
    }

    func restore(_ id: String) async throws {
        try await fire("POST", "api/media/\(id)/restore")
    }

    func archive(_ id: String) async throws {
        try await fire("POST", "api/media/\(id)/archive")
    }

    /// Returns the server's new `edit_version` so the caller can cache-bust
    /// the (immutably cached) versioned thumb URL and stay in step with what
    /// a refresh would return.
    func rotate(_ id: String, degrees: Int = 90) async throws -> Int? {
        let resp = try await fire("POST", "api/media/\(id)/rotate", ["degrees": String(degrees)])
        return resp["edit_version"]?.intValue
    }

    func share(_ id: String) async throws -> ShareResp {
        try await send(try request("POST", "api/media/\(id)/share"), as: ShareResp.self)
    }

    func edit(_ id: String, params: [String: JSONValue]) async throws {
        try await fire("POST", "api/media/\(id)/edit", body: try Self.encoder.encode(params))
    }

    func createUserAlbum(name: String, mediaIds: [String] = []) async throws -> UserAlbumResp {
        try await send(
            try request("POST", "api/user_albums", body: try Self.encoder.encode(NewAlbumBody(name: name, mediaIds: mediaIds))),
            as: UserAlbumResp.self)
    }

    func addToUserAlbum(_ id: Int, ids: [String]) async throws -> UserAlbumResp {
        try await send(
            try request("POST", "api/user_albums/\(id)/items", body: try Self.encoder.encode(IdsBody(ids: ids))),
            as: UserAlbumResp.self)
    }

    func removeFromUserAlbum(_ id: Int, ids: [String]) async throws -> UserAlbumResp {
        try await send(
            try request("DELETE", "api/user_albums/\(id)/items", body: try Self.encoder.encode(IdsBody(ids: ids))),
            as: UserAlbumResp.self)
    }

    func deleteUserAlbum(_ id: Int) async throws {
        _ = try await send(try request("DELETE", "api/user_albums/\(id)"), as: OkResp.self)
    }

    func shareUserAlbum(_ id: Int) async throws -> AlbumShareResp {
        try await send(try request("POST", "api/user_albums/\(id)/share"), as: AlbumShareResp.self)
    }
}

/// Just enough of a dynamic JSON value to read `edit_version` back out of a
/// write response and to send the editor's mixed number/bool parameters.
enum JSONValue: Codable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null
    case array([JSONValue])
    case object([String: JSONValue])

    var intValue: Int? {
        if case .number(let d) = self { return Int(d) }
        return nil
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null }
        else if let b = try? c.decode(Bool.self) { self = .bool(b) }
        else if let d = try? c.decode(Double.self) { self = .number(d) }
        else if let s = try? c.decode(String.self) { self = .string(s) }
        else if let a = try? c.decode([JSONValue].self) { self = .array(a) }
        else if let o = try? c.decode([String: JSONValue].self) { self = .object(o) }
        else { self = .null }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let d): try c.encode(d)
        case .bool(let b): try c.encode(b)
        case .null: try c.encodeNil()
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        }
    }
}
